### PR TITLE
Use pexpect when testing the building of snaps

### PR DESCRIPTION
**Pull request by elopio**

### CODE REVIEW

**Comment by elopio** [comment]
https://bugs.launchpad.net/snapcraft/+bug/1592943

File: snaps_tests/__init__.py [hunk 1]
```diff
@@ -96,6 +111,7 @@ def setUp(self):
                 self.skipTest(
                     '{} does not match the filter {}'.format(
                         self.snap_content_dir, filter_))
+        logger.info('Testing ' + self.snap_content_dir)
```
Comment (sergiusens) [inline snaps_tests/__init__.py]: What happened to `.format` ;-)

File: snaps_tests/__init__.py [hunk 2]
```diff
@@ -138,15 +154,41 @@ def _set_up_qemu_testbed(self):
...
-                stderr=subprocess.STDOUT)
-        except subprocess.CalledProcessError as e:
-            self.addDetail('output', content.text_content(str(e.output)))
-            raise
+            process.expect(expect)
+        except pexpect.ExceptionPexpect:
+            self._add_output_detail(process.before)
+            raise CommandError(
+                command, working_dir,
+                'Expected output "{}" not found.'.format(expect)) from None
```
Comment (sergiusens) [inline snaps_tests/__init__.py]: Use `{!r}` instead and no need for `"` then

File: snaps_tests/__init__.py [hunk 3]
```diff
@@ -138,15 +154,41 @@ def _set_up_qemu_testbed(self):
...
+        finally:
+            process.close()
+            if process.exitstatus:
+                self._add_output_detail(process.before)
+                raise CommandError(
+                    command, working_dir,
+                    'Exit status: {}.'.format(process.exitstatus))
```
Comment (sergiusens) [inline snaps_tests/__init__.py]: {!r}

File: snaps_tests/__init__.py [hunk 4]
```diff
@@ -138,15 +154,41 @@ def _set_up_qemu_testbed(self):
...
+
+    def _add_output_detail(self, output):
+        self.addDetail('output', content.text_content(str(output)))
```
Comment (sergiusens) [inline snaps_tests/__init__.py]: shouldn't this be decoded instead of just `str`ed?

**Comment by sergiusens** [comment]
Looks good, just some inconcistencies with missing use of `{!r}` (I didn't get to all of them) and string concat without using `format` which I thought was our prefered method

File: snaps_tests/__init__.py (hunk 4 above)
Comment (elopio) [inline snaps_tests/__init__.py]: I think it doesn't even need the str. pexpect's before returns a string, I added it to be extra safe.

**Comment by elopio** [comment]
retest this please

**Comment by sergiusens** [comment]
:+1:

**Pull request merged**